### ISSUE TITLE
Buy a developer certificate for bundle signing

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#3341]
## 🧰 Task
To get rid of the scary "cannot verify developer" messages described in the [bundle install guide](https://napari.org/tutorials/fundamentals/installation#installing-as-a-bundled-app)

See #496, #1289, #1074, #3339

**Comment by psobolewskiPhD** [comment]
I think on Mac side this shouldn't cost anything?
https://developer.apple.com/support/membership-fee-waiver/

**Comment by sofroniewn** [comment]
I can get the certification (cc @jni) and then pair with @goanpeca and @jaimergp to get the app signed appropriately

**Comment by Czaki** [comment]
It will be nice to find such certificate which could be used on CI. Some time ago this easiest to reach allow to sign only locally. Which may introduce additional maintenance job.

**Comment by potating-potato** [comment]
Update: I finally got my hand on a certificate, however, the app isn't signed properly yet:
I added the code signing in my [test workflow](https://github.com/potating-potato/napari-demo/blob/main/.github/workflows/test_and_deploy.yml) and generated a [build](https://github.com/potating-potato/napari/suites/4261689944/artifacts/110727563) with the signature. However, the not-verified dialog still shows up when I try to install this bundle. And some debug info from spctl:
```
zliu@CZIMACOS3562 Downloads % spctl --assess -v /Applications/napari.app 
/Applications/napari.app: rejected
source=Unnotarized Developer ID
```

So I did a bit more research to notarize the app if we want to [distribute outside of apple app store](https://help.apple.com/xcode/mac/current/#/dev033e997ca). I tried to upload the bundle and saw the following [notarization failure](https://osxapps-ssl.itunes.apple.com/itunes-assets/Enigma126/v4/f7/85/1f/f7851f16-f88d-907d-748c-5b6ff1d98020/developer_log.json?accessKey=1636289544_4380334595618108098_P2fUK7Jdj5Yn%2Fapn1jb9ZvX5%2Bz4OQxks4D6dKHY1oFAlGEv05EykTGE740JXDk8YwuwadtZe1syqNE9nos%2F4Nk2I8TIQ9qNIo5%2F8H9jDBlsrLnJP6sYTeShMGJYi2toRD%2FsIBnvk%2BhYkWxA7qxI3qv7nHlctTD9Dg4YVgXuTyBo%3D)

I am not sure if notarization is absolutely required from the wording of it. I will continue diving into this, but any advice is appreciated

**Comment by potating-potato** [comment]
some more info on the signature:
```zliu@CZIMACOS3562 Downloads % spctl -a -v /Applications/napari.app
/Applications/napari.app: rejected
source=Unnotarized Developer ID
zliu@CZIMACOS3562 Downloads % codesign -dv --verbose=4 /Applications/napari.app 
Executable=/Applications/napari.app/Contents/MacOS/napari
Identifier=com.napari.napari
Format=app bundle with Mach-O universal (x86_64 arm64)
CodeDirectory v=20500 size=76701 flags=0x10000(runtime) hashes=2388+5 location=embedded
VersionPlatform=1
VersionMin=720896
VersionSDK=721664
Hash type=sha256 size=32
CandidateCDHash sha256=8656493b4c031f5c4ddbb78cf12f4d8a0c2933ed
CandidateCDHashFull sha256=8656493b4c031f5c4ddbb78cf12f4d8a0c2933ed3687d0e490e5de4aab192fc0
Hash choices=sha256
CMSDigest=8656493b4c031f5c4ddbb78cf12f4d8a0c2933ed3687d0e490e5de4aab192fc0
CMSDigestType=2
Page size=4096
CDHash=8656493b4c031f5c4ddbb78cf12f4d8a0c2933ed
Signature size=9012
Authority=Developer ID Application: Chan Zuckerberg Initiative, LLC (7WAT66VT96)
Authority=Developer ID Certification Authority
Authority=Apple Root CA
Timestamp=Nov 4, 2021 at 3:15:30 PM
Info.plist entries=13
TeamIdentifier=7WAT66VT96
Runtime Version=11.3.0
Sealed Resources version=2 rules=13 files=29241
Internal requirements count=1 size=180

zliu@CZIMACOS3562 Downloads % spctl -a -v /Applications/napari.app      
/Applications/napari.app: rejected
source=Unnotarized Developer ID
```

**Comment by Czaki** [comment]
looking at the log, one of the problems for apple is that app files contain many unsigned binaries, even if they are exported outside `.app`.  The question is that even if they are properly signed by napari certificate what will happen if the user installs a plugin that provides its own executable (or one of its dependencies).  If this will break signature verification.

**Comment by potating-potato** [comment]
yeah. from what I read I am skeptical that we will ever pass apple notarization.

Currently, I am puzzled if the notarization step is absolutely necessary for user to open the app without triggering the gatekeeper blocking, from the wording on the website it seems as long as the app is signed it is fine. But from my run it is then requiring notarization...

**Comment by jaimergp** [comment]
I _think_ conda-forge packages are signed in osx, so maaaybe there's a chance we could (notice all the conditionals) pass it when we switch to conda-based bundles.

**Comment by Czaki** [comment]
> I _think_ conda-forge packages are signed in osx, so maaaybe there's a chance we could (notice all the conditionals) pass it when we switch to conda-based bundles.

There is still a question of what will happen if the user installs some plugin from pip.

**Comment by jaimergp** [comment]
Those are post-bundle and I think we will have them in a separate environment so hopefully that's ok.

**Comment by potating-potato** [comment]
ok for now I will stop looking into this, I have the certificate ready when you guys are ready to test conda based bundles

**Comment by jaimergp** [comment]
> I think conda-forge packages are signed in osx

Correction, they are self-signed at install time 😬 Let's see if notarization helps, otherwise we might need to codesign (not installer-sign) everything inside the PKG.

**Comment by potating-potato** [comment]
Update on this: we are able to sign our conda MacOS bundle with a certificate for now to get rid of the security message, we are still working on signing identity and windows certificate before we can close this issue.

**Comment by jaimergp** [comment]
To be clear, Windows installers are also signed, technically, _but_ using the same Apple certificate we use for PKGs, which Microsoft does not recognize as "good enough" to disable the SmartScreen warning. That said, if a user wants to inspect the signature, they'll see that it's using the napari certificate.
